File: Data Structures & Algorithms/house-robber/submission-1.java
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 1) return nums[0];
        if (nums.length == 2) return Math.max(nums[0], nums[1]);
        int[] suffixDP = new int[nums.length];
        suffixDP[nums.length-1] = nums[nums.length-1];
        suffixDP[nums.length-2] = nums[nums.length-2];
        for (int i=nums.length-3; i>=0; i--) {
            suffixDP[i] = Math.max(suffixDP[i+1], nums[i]+suffixDP[i+2]);
        }
        return suffixDP[0];
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root) {
        if(!root) return 0;

        let max = 0;

        const getDiameter = (root) => {
            if(!root) return 0;

            let left = getHeight(root.left);
            let right = getHeight(root.right);
  
            max = Math.max(max, left + right, getDiameter(root.left), getDiameter(root.right));
            return max;
        }

        const getHeight = (root) => {
            if(!root) return 0;

            return Math.max(getHeight(root.left), getHeight(root.right)) + 1;
        }

        return getDiameter(root);
    }
}
